src/wallet/wallet2.cpp

```diff
@@ -5326,51 +10509,31 @@ uint64_t wallet2::get_daemon_blockchain_height(string &err)
 
 uint64_t wallet2::get_daemon_blockchain_target_height(string &err)
 {
-  epee::json_rpc::request<cryptonote::COMMAND_RPC_GET_INFO::request> req_t = AUTO_VAL_INIT(req_t);
-  epee::json_rpc::response<cryptonote::COMMAND_RPC_GET_INFO::response, std::string> resp_t = AUTO_VAL_INIT(resp_t);
-  m_daemon_rpc_mutex.lock();
-  req_t.jsonrpc = "2.0";
-  req_t.id = epee::serialization::storage_entry(0);
-  req_t.method = "get_info";
-  bool ok = net_utils::invoke_http_json("/json_rpc", req_t, resp_t, m_http_client);
-  m_daemon_rpc_mutex.unlock();
-  if (ok)
-  {
-    if (resp_t.result.status == CORE_RPC_STATUS_BUSY)
-    {
-      err = "daemon is busy. Please try again later.";
-    }
-    else if (resp_t.result.status != CORE_RPC_STATUS_OK)
-    {
-      err = resp_t.result.status;
-    }
-    else // success, cleaning up error message
-    {
-      err = "";
-    }
-  }
-  else
+  err = "";
+  uint64_t target_height = 0;
+  const auto result = m_node_rpc_proxy.get_target_height(target_height);
+  if (result && *result != CORE_RPC_STATUS_OK)
   {
-    err = "possibly lost connection to daemon";
+    err= *result;
+    return 0;
   }
-  return resp_t.result.target_height;
+  return target_height;
 }
 
 uint64_t wallet2::get_approximate_blockchain_height() const
 {
-  // time of begining: testnet: 2018-01-12, mainnet: 2018-01-18;
-  const time_t fork_time = m_testnet ? 1515715200 : 1516233600;
-
-  // in case we not launched mainnet yet
-  if (fork_time > time(nullptr))
-    return 0;
-
-  // v2 fork block
-  const uint64_t fork_block = m_testnet ? 1 : 1;
+  // TODO: graft stagenet and next fork block and time
+  const time_t fork_time = m_nettype == TESTNET ? 1515715200 : m_nettype == STAGENET ? (time_t)-1/*TODO*/ : 1516233600;
+  // v7 fork block
+  const uint64_t fork_block = m_nettype == TESTNET ? 1 : m_nettype == STAGENET ? (uint64_t)-1/*TODO*/ : 1;
   // avg seconds per block
   const int seconds_per_block = DIFFICULTY_TARGET_V2;
   // Calculated blockchain height
   uint64_t approx_blockchain_height = fork_block + (time(NULL) - fork_time)/seconds_per_block;
+  // testnet got some huge rollbacks, so the estimation is way off
+  static const uint64_t approximate_testnet_rolled_back_blocks = 303967;
```

Comment:
Monero constant, probably cannot be applied for Graft

Reply:
removed